File: apps/ios/Sources/CrossyUI/PuzzlesScreen.swift
// Puzzles, the library tab: the caller's uploaded puzzles from GET /puzzles, newest
// first, cursor-paginated, the RoomsScreen grammar exactly (title in the scroll,
// paper cards, pull to refresh, quiet state lines). Each card's one action starts a
// fresh game from that upload (POST /games, the replay-without-reupload path; the web
// gallery mirrors it): on success the composition root pushes the created room, the
// same navigation an opened room card takes.
//
// The screen owns scroll, paging, states, and per-card start progress; it reaches the
// network only through the two injected closures (AD-2: CrossyUI sees neither
// CrossyAPI nor the protocol twins, the composition root adapts both). A start
// resolves to the created gameId or an ArrivalFailure; success is handed up so the
// root does the push, a failure reads inline on the card and the card recovers.

import CrossyDesign
import SwiftUI

/// One page of puzzles as the screen consumes it (the RoomsPage pattern): rows plus
/// the `before` cursor for the next page, nil when this page ended the list
/// (PROTOCOL.md §12 pagination, already digested by the composition root).
public struct PuzzlesPage: Equatable, Sendable {
    public let puzzles: [PuzzleCardModel]
    public let nextBefore: String?

    public init(puzzles: [PuzzleCardModel], nextBefore: String?) {
        self.puzzles = puzzles
        self.nextBefore = nextBefore
    }
}

public struct PuzzlesScreen: View {
    private let loadPage: (String?) async -> Result<PuzzlesPage, ArrivalFailure>
    /// Start a game from a puzzle: `POST /games` behind the seam, resolving to the
    /// created gameId on success. The screen owns the in-flight and failure state;
    /// the root does the navigation with the returned id.
    private let startGame: (PuzzleCardModel) async -> Result<String, ArrivalFailure>
    /// The created room, handed to the composition root to push (the RoomsScreen
    /// onOpenRoom shape: navigation is the root's, not the screen's).
    private let onOpenRoom: (String) -> Void
    /// The puzzle a crossy://play deep link preselects (D22, the web's `?play=`): scroll to
    /// it and briefly ring it so the solver sees which upload to start. The start stays a
    /// manual tap; puzzle import never auto-creates a game.
    private let playIntent: String?

    @Environment(\.colorScheme) private var colorScheme
    @State private var puzzles: [PuzzleCardModel] = []
    @State private var nextBefore: String?
    @State private var exhausted = false
    @State private var loaded = false
    @State private var loadingMore = false
    @State private var failure: ArrivalFailure?
    /// The puzzleId whose `POST /games` is currently out; nil when none. One start at
    /// a time is enough (a person taps one card), and it keeps the disabled control
    /// unambiguous.
    @State private var starting: String?
    /// A per-card inline failure sentence, keyed by puzzleId; cleared when that card
    /// is tapped again.
    @State private var startFailures: [String: String] = [:]
    /// The card currently ringed by a play intent, cleared after a beat; nil when none.
    @State private var highlighted: String?
    /// The play intent already honored, so a re-render never re-scrolls (the web's
    /// consume-once), keyed by the intent's puzzleId.
    @State private var consumedIntent: String?

    public init(
        loadPage: @escaping (String?) async -> Result<PuzzlesPage, ArrivalFailure>,
        startGame: @escaping (PuzzleCardModel) async -> Result<String, ArrivalFailure> = { _ in
            .failure(ArrivalFailure(code: nil))
        },
        onOpenRoom: @escaping (String) -> Void = { _ in },
        playIntent: String? = nil
    ) {
        self.loadPage = loadPage
        self.startGame = startGame
        self.onOpenRoom = onOpenRoom
        self.playIntent = playIntent
    }

    private var ground: GridGround {
        colorScheme == .dark ? .observatory : .studio
    }

    public var body: some View {
        ScrollViewReader { proxy in
            ScrollView {
                LazyVStack(alignment: .leading, spacing: 10) {
                    Text(verbatim: ArrivalCopy.puzzlesTitle)
                        .font(.system(size: 32, weight: .bold))
                        .foregroundStyle(Color(rgb: ground.tokens.ink))
                        .padding(.top, 8)
                        .padding(.bottom, 8)

                    if let failure, puzzles.isEmpty {
                        stateLine(failure.sentence)
                    } else if loaded && puzzles.isEmpty {
                        stateLine(ArrivalCopy.puzzlesEmpty)
                    }

                    ForEach(puzzles) { puzzle in
                        PuzzleCard(
                            model: puzzle,
                            ground: ground,
                            starting: starting == puzzle.id,
                            failure: startFailures[puzzle.id],
                            onStart: { Task { await start(puzzle) } }
                        )
                        .id(puzzle.id)
                        .overlay {
                            if highlighted == puzzle.id {
                                RoundedRectangle(cornerRadius: 16, style: .continuous)
                                    .strokeBorder(
                                        Color(rgb: ground.tokens.ink), lineWidth: 2)
                                    .transition(.opacity)
                            }
                        }
                        .onAppear {
                            if puzzle.id == puzzles.last?.id {
                                Task { await loadMore() }
                            }
                        }
                    }
                }
                .padding(.horizontal, 16)
                .padding(.bottom, 16)
            }
            .refreshable { await reload() }
            .background(Color(rgb: ground.tokens.canvas).ignoresSafeArea())
            .task {
                await reload()
                await honorIntent(proxy)
            }
            .onChange(of: playIntent) { Task { await honorIntent(proxy) } }
        }
    }

    /// Honor a crossy://play preselect once: land, scroll the named upload to center and
    /// ring it briefly so the solver sees which one to start. Reloads first when the card is
    /// missing (a just-imported puzzle is not in a cached list); an id this account does not
    /// have (foreign, deleted, stale) rings nothing and leaves the plain library.
    private func honorIntent(_ proxy: ScrollViewProxy) async {
        guard let intent = playIntent, intent != consumedIntent else { return }
        consumedIntent = intent
        if !puzzles.contains(where: { $0.id == intent }) { await reload() }
        guard puzzles.contains(where: { $0.id == intent }) else { return }
        withAnimation(.easeInOut) { proxy.scrollTo(intent, anchor: .center) }
        withAnimation { highlighted = intent }
        try? await Task.sleep(for: .seconds(1.8))
        if highlighted == intent { withAnimation { highlighted = nil } }
    }

    private func stateLine(_ sentence: String) -> some View {
        Text(verbatim: sentence)
            .font(.system(size: 14))
            .foregroundStyle(Color(rgb: ground.tokens.number))
            .frame(maxWidth: .infinity, alignment: .center)
            .multilineTextAlignment(.center)
            .padding(.top, 48)
    }

    // MARK: - Start a game (POST /games; the create-flow slice, closed)

    /// Start a fresh game from one puzzle: mark the card in flight, clear any prior
    /// failure on it, then run the injected create. Success hands the created gameId
    /// to the root, which pushes the room the same way an opened card does; a failure
    /// stays on the list and reads inline on the card (a toast would be noise), keyed
    /// on the §12 code. A second tap while one start is out is a no-op (the button is
    /// already disabled, this guards the closure too).
    private func start(_ puzzle: PuzzleCardModel) async {
        guard starting == nil else { return }
        starting = puzzle.id
        startFailures[puzzle.id] = nil
        defer { starting = nil }
        switch await startGame(puzzle) {
        case .success(let gameId):
            onOpenRoom(gameId)
        case .failure(let arrivalFailure):
            startFailures[puzzle.id] = ArrivalCopy.puzzleStartFailure(
                forCode: arrivalFailure.code)
        }
    }

    // MARK: - Paging (the RoomsScreen contract, verbatim)

    private func reload() async {
        failure = nil
        switch await loadPage(nil) {
        case .success(let page):
            puzzles = page.puzzles
            nextBefore = page.nextBefore
            exhausted = page.nextBefore == nil
        case .failure(let arrivalFailure):
            failure = arrivalFailure
        }
        loaded = true
    }

    private func loadMore() async {
        guard !exhausted, !loadingMore, let cursor = nextBefore else { return }
        loadingMore = true
        defer { loadingMore = false }
        switch await loadPage(cursor) {
        case .success(let page):
            // An empty page is the §12 end of iteration; appending nothing and
            // stopping is exactly the cursor contract.
            puzzles.append(contentsOf: page.puzzles)
            nextBefore = page.nextBefore
            exhausted = page.puzzles.isEmpty || page.nextBefore == nil
        case .failure:
            // A failed load-more never blanks what is already on screen; pull to
            // refresh is the recovery.
            exhausted = false
        }
    }
}

File: apps/android/ui/src/main/kotlin/crossy/ui/PuzzlesScreen.kt
// Puzzles, the library tab (iOS PuzzlesScreen): the caller's uploaded puzzles from GET /puzzles,
// newest first, cursor-paginated, the RoomsListScreen grammar exactly (a big title in the scroll,
// paper cards, pull to refresh, quiet state lines). Each card's one action starts a fresh game from
// that upload (POST /games, the replay-without-reupload path the web gallery mirrors): on success
// the composition root pushes the created room, the same navigation an opened room card takes; a
// failure reads inline on the card and the card recovers. Android keeps a manual "Create room" entry
// too (the named-create + id-paste screen, a deliberate extra), reached from the header here.
//
// A pure function of the loaded list plus its loading/refreshing state and the per-card start
// progress; the composition root owns the fetch, the paging, the start call, and the navigation.

package crossy.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import crossy.protocol.PuzzleSummary

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PuzzlesScreen(
    puzzles: List<PuzzleSummary>,
    ground: GridGround,
    isLoading: Boolean,
    isRefreshing: Boolean,
    error: String?,
    // The puzzleId whose POST /games is currently out; null when none. One start at a time is enough
    // (a person taps one card) and it keeps the disabled control unambiguous (iOS PuzzlesScreen.starting).
    startingId: String?,
    // A per-card inline failure sentence, keyed by puzzleId; the host clears it when that card is
    // tapped again (iOS PuzzlesScreen.startFailures).
    startFailures: Map<String, String>,
    onStart: (PuzzleSummary) -> Unit,
    onRefresh: () -> Unit,
    onLoadMore: () -> Unit,
    onCreate: () -> Unit,
) {
    val lastId = puzzles.lastOrNull()?.puzzleId
    PullToRefreshBox(
        isRefreshing = isRefreshing,
        onRefresh = onRefresh,
        modifier = Modifier.fillMaxSize(),
    ) {
        LazyColumn(
            modifier = Modifier.fillMaxSize().padding(horizontal = 16.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            item(key = "header") {
                Column {
                    Row(
                        modifier = Modifier.fillMaxWidth().padding(top = 8.dp, bottom = 8.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(
                            ArrivalCopy.puzzlesTitle,
                            fontSize = 32.sp,
                            fontWeight = FontWeight.Bold,
                            color = MaterialTheme.colorScheme.onSurface,
                            modifier = Modifier.weight(1f),
                        )
                        // Android's deliberate extra: the named-create + id-paste screen, reachable
                        // here so a room can be created by hand even when the upload list is empty.
                        TextButton(onClick = onCreate) { Text("Create room") }
                    }
                    if (error != null && puzzles.isEmpty()) {
                        Text(
                            error,
                            color = MaterialTheme.colorScheme.error,
                            fontSize = 13.sp,
                            modifier = Modifier.padding(bottom = 8.dp),
                        )
                    }
                }
            }

            when {
                isLoading && puzzles.isEmpty() -> item(key = "loading") {
                    Box(Modifier.fillMaxWidth().padding(top = 32.dp), contentAlignment = Alignment.Center) {
                        CircularProgressIndicator()
                    }
                }
                puzzles.isEmpty() -> item(key = "empty") {
                    Text(
                        ArrivalCopy.puzzlesEmpty,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        fontSize = 14.sp,
                        modifier = Modifier.padding(vertical = 16.dp),
                    )
                }
                else -> items(puzzles, key = { it.puzzleId }) { puzzle ->
                    LibraryPuzzleCard(
                        puzzle = puzzle,
                        ground = ground,
                        starting = startingId == puzzle.puzzleId,
                        failure = startFailures[puzzle.puzzleId],
                        onStart = { onStart(puzzle) },
                    )
                    if (puzzle.puzzleId == lastId) {
                        LaunchedEffect(lastId) { onLoadMore() }
                    }
                }
            }
        }
    }
}

/** One library puzzle card (iOS PuzzleCard): geometry fingerprint, title, author, and the one action
 *  that starts a fresh game from this upload (POST /games). No people row (a puzzle has no members,
 *  and people are the only color a card earns), so the start action is the card's only tint-free
 *  control. In flight it reads "Starting" and disables, so a double tap never fires a second create.
 *  A per-card failure line reads beneath, inline, no toast. */
@Composable
private fun LibraryPuzzleCard(
    puzzle: PuzzleSummary,
    ground: GridGround,
    starting: Boolean,
    failure: String?,
    onStart: () -> Unit,
) {
    androidx.compose.material3.Card(modifier = Modifier.fillMaxWidth()) {
        Column(Modifier.fillMaxWidth().padding(14.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(14.dp),
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth(),
            ) {
                GeometryFingerprint(puzzle.rows, puzzle.cols, ground, Modifier.size(52.dp))
                Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(4.dp)) {
                    Text(
                        puzzle.headline,
                        fontSize = 16.sp,
                        fontWeight = FontWeight.SemiBold,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        color = MaterialTheme.colorScheme.onSurface,
                    )
                    puzzle.subline?.let {
                        Text(
                            it,
                            fontSize = 13.sp,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }
                Button(onClick = onStart, enabled = !starting) {
                    Text(if (starting) ArrivalCopy.puzzleStarting else ArrivalCopy.puzzleStartGame)
                }
            }
            failure?.let {
                Text(it, fontSize = 13.sp, color = MaterialTheme.colorScheme.error)
            }
        }
    }
}

/** The headline: the title when the document carried one, else the honest geometry (the RoomCard
 *  fallback, same words). */
private val PuzzleSummary.headline: String
    get() = title?.takeIf { it.isNotBlank() } ?: "${rows}×${cols} crossword"

/** The author under the title, absent rather than empty when the document carried none. */
private val PuzzleSummary.subline: String?
    get() = author?.takeIf { it.isNotBlank() }

/** A bare geometry lattice (iOS GeometryFingerprintView): a puzzle summary carries no mask (INV-6,
 *  and there is nothing to solve yet), so the card draws only the grid's shape, `rows` by `cols`,
 *  cell-toned with quiet interior lines. Geometry only, so INV-6 holds structurally. */
@Composable
private fun GeometryFingerprint(rows: Int, cols: Int, ground: GridGround, modifier: Modifier = Modifier) {
    val cell = ground.tokens.cell.toColor()
    val line = ground.tokens.gridLine.toColor()
    Canvas(modifier = modifier.clip(RoundedCornerShape(6.dp))) {
        if (rows <= 0 || cols <= 0) return@Canvas
        val step = minOf(size.width / cols, size.height / rows)
        val w = step * cols
        val h = step * rows
        drawRect(color = cell, topLeft = Offset(0f, 0f), size = androidx.compose.ui.geometry.Size(w, h))
        for (c in 0..cols) {
            drawLine(line, Offset(c * step, 0f), Offset(c * step, h), strokeWidth = 1f)
        }
        for (r in 0..rows) {
            drawLine(line, Offset(0f, r * step), Offset(w, r * step), strokeWidth = 1f)
        }
    }
}
